sdubov/aws-toolkit-jetbrains
# Enable backend gutter marks for lambda functions

## Types of changes

## Description
- Add AWS Rider backend C# logic
- Update gradle to compile AWS Rider backend logic and copy it into plugin output
- Add Rider Protocol extensions to handle backend events
- Update gradle to generate protocol models on Kotlin and C# sides
- Add fake backend icons for actions in Lambda gutter mark context
- Add Icons patcher on frontend to update fake backend icons with frontend icons
- Pass credentials when opening run lambda run configuration
- Renable the Configuration type instance to separate from configuration type
- Minor updates

## Motivation and Context
Gutter marks allow users to run/debug lambda from editor from a gutter context menu. To set a gutter mark to a correct lambda method we should parse PSI to determine a right function that represents a lambda. In Rider all information about PSI lives on a backend. To make a gutter mark work in Rider, we should enable a backend in plugin and show a backend gutter mark with backend actions.

Since we have all actions defined on a frontend we should fire event on user interactions with context icons that are defined on a backend. This is enabled by Rider protocol extension that is listening for a backend gutter mark popup menu event and fire action on a frontend.

## Related Issue(s)
None

## Testing
```
JetBrains Rider 2019.2 EAP
Runtime version: 1.8.0_212-release-1566-b8 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.5
GC: PS Scavenge, PS MarkSweep
Memory: 480M
Cores: 8
```
- Run/debug/create actions from a gutter mark in Rider
- Verify new Gradle tasks
- Verify Idea plugin gutter marks

## Screenshots (if appropriate)
<img width="652" alt="LambdaGutterMark" src="https://user-images.githubusercontent.com/6064345/61903224-ad27cf80-af2c-11e9-92b8-257d171defc2.png">

## Checklist

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.
